rootfs/Dockerfile

```diff
@@ -12,7 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM alpine:3.7
+FROM BASEIMAGE
+
+CROSS_BUILD_COPY qemu-QEMUARCH-static /usr/bin/
```

Comment:
I'm confused here. This is the docker file for the released image isn't it? It should not contain an emulator, as the image should run natively on the target arch.

In general, I'm also a bit confused by including an emulator at all. I thought golang could cross compile all by itself. Greping through the kubernetes code for qemu, they do have qemu in there, but from a superficial skimming of the code, only looks to be used to test the binaries, not build/ship them? Maybe I'm totally wrong.

Reply:
Hi @kfox1111 
Thank you for your comments. :)

I have added a command to delete the emulator in the end of each Dockerfile.

Regarding to your second question, you should notice the following line in each Dockerfile:
"RUN apk update && apk add ... "
It means, it needs to execute the arm instruction in docker building step. 
So the emulator in Dockerfile is mandatory.
Regarding to rook/ceph, we don't need to copy emulator in Dockerfile. Because we don't need to run any non-x86 instruction in docker building step.
But for rook/nfs, the emulator in Dockerfile is mandatory. The reason for this problem is same as the problem in helm.
You can see that there are many commands in the Dockerfile of ‘rook/nfs' that need to use the non-x86 instruction. Such as: 'dnf install', 'make' ...

You can test my patch in your environment.
The building will be failed if without the following line in Dockerfile:
"CROSS_BUILD_COPY qemu-QEMUARCH-static /usr/bin/"
The error should be like following:

Step 2/9 : RUN apk update && apk add ca-certificates socat && rm -rf /var/cache/apk/*
 ---> Running in 02f412954455
standard_init_linux.go:190: exec user process caused "no such file or directory"
The command '/bin/sh -c apk update && apk add ca-certificates socat && rm -rf /var/cache/apk/*' returned a non-zero code: 1
Makefile:120: recipe for target 'docker-build' failed

Introduction article: https://eyskens.me/multiarch-docker-images/

In addition, I need to explain that these problems have nothing to do with golang cross compilation.
For helm, the problem is that it uses the following command in Dockerfile:
"RUN apk update && apk add ... "

For rook/nfs, the problem is that it uses the following command in Dockerfile:
"RUN dnf install ... make...."

Reply:
Hmm... I see. thanks for the explanation.

removing the emulator in a subsequent layer will only hide it, not remove it. so it will still be shipped to the end devices anyway.

I was thinking the emulator was fairly sizeable, but it looks like its pretty small compared to tiller so it may not be too bad to ship to machines that never will use it.

Reply:
This is a tricky problem @justincormack might have an idea other than the experimental version of Docker? In OpenFaaS we build for amd64, armhf, arm64 but with the two arm platforms built on a device and not via cloud CI. It's a huge pain so we may move to qemu like this.

Reply:
You can remove the emulator using a multi-stage build. Cross builds are kind of horrible compared to native, but better than no build...